sw/host/tests/crypto/aes_gcm_nist_kat/src/main.rs

```diff
@@ -0,0 +1,231 @@
+        test_case.aad.len(),
+        AES_GCM_CMD_MAX_AAD_BYTES,
+    );
+    assert!(
+        test_case.tag.len() <= AES_GCM_CMD_MAX_TAG_BYTES,
+        "Tag too long for device firmware configuration (got = {}, max = {})",
+        test_case.tag.len(),
+        AES_GCM_CMD_MAX_TAG_BYTES,
+    );
+
+    // Send operation type
+    match test_case.operation.as_str() {
+        "encrypt" => CryptotestAesGcmOperation::Encrypt,
+        "decrypt" => CryptotestAesGcmOperation::Decrypt,
+        _ => panic!("Unsupported operation type"),
+    }
+    .send(&*uart)?;
+
+    // Set input data
+    let input_data = match test_case.operation.as_str() {
+        "encrypt" => &test_case.plaintext,
+        "decrypt" => &test_case.ciphertext,
+        _ => panic!("Invalid operation type")
+    };
+
+    // Send data
+    CryptotestAesGcmData {
+        key: ArrayVec::try_from(test_case.key.as_slice()).unwrap(),
+        key_length: test_case.key.len(),
+        iv: ArrayVec::try_from(test_case.iv.as_slice()).unwrap(),
+        iv_length: test_case.iv.len(),
+        aad: ArrayVec::try_from(test_case.aad.as_slice()).unwrap(),
+        aad_length: test_case.aad.len(),
+        tag: ArrayVec::try_from(test_case.tag.as_slice()).unwrap(),
+        tag_length: test_case.tag.len(),
+        input: ArrayVec::try_from(input_data.as_slice()).unwrap(),
+        input_length: input_data.len(),
+    }
+    .send(&*uart)?;
+
+    // Get the output
+    let gcm_output = CryptotestAesGcmOutput::recv(&*uart, opts.timeout, false)?;
+
+    let mut failed = false;
+    vec![("oneshot", gcm_output.oneshot_output),
+         ("stepwise", gcm_output.stepwise_output)]
+    .into_iter()
+    .for_each(|(mode, output)|{
+        let target_output = match test_case.operation.as_str() {
+            "encrypt" => &test_case.ciphertext,
+            "decrypt" => &test_case.plaintext,
+            _ => panic!("Invalid operation type")
+        };
+        let success = if target_output.len() > gcm_output.output_length {
+            // If we got a shorter digest back then the test asks for, we
+            // can't accept the digest, even if the beginning bytes match.
+            false
+        } else {
+            // Some test cases only specify the beginning bytes of the
+            // expected digest, so we only check up to what the test
+            // specifies.
+            target_output[..] == output[..target_output.len()]
+        };
+        if target_output.len() > 0 && test_case.result != success {
+            log::info!(
+                "FAILED {} test #{} in {} mode: expected = {}, actual = {}",
+                test_case.operation,
+                test_case.test_case_id,
+                mode,
+                test_case.result,
+                success
+            );
+            failed = true;
+        }
+    });
+    if failed {
+        *fail_counter += 1;
+    }
+    Ok(())
+}
+
+fn is_input_empty(aes_gcm_test: &AesGcmTestCase) -> bool {
```

Comment:
What's the purpose of this function? Does cryptolib not accept empty inputs? If that's the case, those tests should be thrown out in the test vector parser, rather than forcing Rust to filter out unused tests (adding a change to the test vector parser to this PR is okay if that's what needs to happen).

Reply:
I need some discussion over it.

`cryptolib` accepts empty inputs, but there are around 10% empty input test vectors out of 48k test vectors. Parsing each test vector roughly takes 0.75 seconds time so I was wondering whether we really should pass them to cryptolib or not.

Reply:
I need some discussion over it.

cryptolib accepts empty inputs, but there are around 10% empty input test vectors out of 48k test vectors. Parsing each test vector roughly takes 0.75 seconds time so I was wondering whether we really should pass them to cryptolib or not.

Reply:
Removing this function at this moment. We can continue our discussion here and if needed we can make desired changes.